core/modules/replica/IngestServer.h

```diff
@@ -0,0 +1,130 @@
+ */
+
+// System headers
+#include <memory>
+
+// Third party headers
+#include <boost/asio.hpp>
+
+// Qserv headers
+#include "replica/Configuration.h"
+#include "replica/IngestServerConnection.h"
+#include "replica/ServiceProvider.h"
+
+// This header declarations
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+/**
+  * Class IngestServer is used for handling incoming connections to
+  * the file delivery service. Each instance of this class will be running
+  * in its own thread.
+  */
+class IngestServer : public std::enable_shared_from_this<IngestServer>  {
+public:
+
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<IngestServer> Ptr;
+
+    /**
+     * Static factory method is needed to prevent issue with the lifespan
+     * and memory management of instances created otherwise (as values or via
+     * low-level pointers).
+     *
+     * @param serviceProvider
+     *   for configuration, etc. services
+     *
+     * @workerName
+     *   the name of a worker this service is acting upon (used for checking
+     *   consistency of the protocol)
+     *
+     * @return
+     *   pointer to the created object
+     */
+    static Ptr create(ServiceProvider::Ptr const& serviceProvider,
+                      std::string const& workerName);
+
+    // Default construction and copy semantics are prohibited
+
+    IngestServer() = delete;
+    IngestServer(IngestServer const&) = delete;
+    IngestServer& operator=(IngestServer const&) = delete;
+
+    ~IngestServer() = default;
+
+    /// @return the name of a worker this server runs for
+    std::string const& worker() const { return _workerName; }
+
+    /**
+     * Run the server in a thread pool (as per the Configuration)
+     *
+     * @note
+     *   This is the blocking operation. Please, run it within its own thread
+     *   if needed.
+     */
+    void run();
+
+private:
+
+    /// @see IngestServer::create()
+    IngestServer(ServiceProvider::Ptr const& serviceProvider,
+                 std::string const& workerName);
+
+    /**
+     * Begin (asynchronously) accepting connection requests.
+     */
+    void _beginAccept();
+    
+    /**
+     * Handle a connection request once it's detected. The rest of
+     * the communication will be forwarded to the connection object
+     * specified as a parameter of the method.
+     */
+    void _handleAccept(IngestServerConnection::Ptr const& connection,
+                       boost::system::error_code const& ec);
+
+    /// @return the context string to be used for the message logging
+    std::string _context() const { return "INGEST-SERVER  "; }
+
+    // Input parameters
+
+    ServiceProvider::Ptr const _serviceProvider;
+    std::string          const _workerName;
+
+    /// Cached worker descriptor obtained from the configuration
+    WorkerInfo const _workerInfo;
+
+    boost::asio::io_service        _io_service;
```

Comment:
For your consideration: I like a pattern where the asio::io_service is passed in at service construction time.  This lets a server implementor (implementing a server that may provide more than one type of service) decide on the appropriate threading model, control which services share and asio::io_service instance and which have their own, gives better control over service lifetime vs. server lifetime, etc.

Reply:
Noted. I will consider this.